File: lib/embed.py
"""
lib/embed.py — Shared embedding utility (Python / pipeline side).

Implements the embedding code path from ATELIER_RETRIEVAL_SPEC.md §4 ("embedding
generation gets one home") and the Layer 0 model lock in §6
(text-embedding-3-small, 1536 dims).

CORRECTNESS RULE (spec §3, non-negotiable): the query is embedded with the SAME
model as the stored chunks — a different model is a different vector space and
silent garbage. EMBEDDING_MODEL below is the single source of truth on the Python
side; lib/embed.ts MUST declare the identical string. Changing the model is a
full re-embed of the table plus a bump of the `model` column, not a config tweak.

Requires:
  - OPENAI_API_KEY in the environment (loaded from .env by config.py / callers)
  - the `openai` package (added to requirements.txt)
"""

from __future__ import annotations

import os
from functools import lru_cache
from typing import List

from openai import OpenAI

# ── Model lock — keep IN SYNC with lib/embed.ts ─────────────────────────────────
EMBEDDING_MODEL = "text-embedding-3-small"
EMBEDDING_DIMENSIONS = 1536


@lru_cache(maxsize=1)
def _client() -> OpenAI:
    api_key = os.environ.get("OPENAI_API_KEY")
    if not api_key:
        raise RuntimeError(
            "OPENAI_API_KEY is not set. Add it to .env before embedding."
        )
    return OpenAI(api_key=api_key)


def embed_texts(texts: List[str]) -> List[List[float]]:
    """
    Embed a batch of strings. Returns one EMBEDDING_DIMENSIONS-length vector per
    input, in the same order as the input list.
    """
    if not texts:
        return []
    # The API rejects empty strings; substitute a single space so output indices
    # stay aligned 1:1 with the caller's input list.
    cleaned = [t if (t and t.strip()) else " " for t in texts]
    resp = _client().embeddings.create(model=EMBEDDING_MODEL, input=cleaned)
    items = sorted(resp.data, key=lambda d: d.index)  # defensive: guarantee order
    vectors = [item.embedding for item in items]
    for v in vectors:
        if len(v) != EMBEDDING_DIMENSIONS:
            raise RuntimeError(
                f"Embedding dimension mismatch: got {len(v)}, "
                f"expected {EMBEDDING_DIMENSIONS} for model {EMBEDDING_MODEL}"
            )
    return vectors


def embed_text(text: str) -> List[float]:
    """Embed a single string. Returns an EMBEDDING_DIMENSIONS-length vector."""
    return embed_texts([text])[0]


def to_pgvector(vector: List[float]) -> str:
    """
    Format a Python float list as a pgvector literal ('[v0,v1,...]') for insertion
    or RPC through PostgREST, which does not accept a raw JSON array for the
    vector type.
    """
    return "[" + ",".join(str(x) for x in vector) + "]"

File: lib/embed.ts
/**
 * lib/embed.ts — Shared embedding utility (web app / query side).
 *
 * Query-side parity for the embedding path in ATELIER_RETRIEVAL_SPEC.md §4,
 * mirroring lib/embed.py. Same model, same dimensions.
 *
 * CORRECTNESS RULE (spec §3, non-negotiable): query and stored chunks must use
 * the SAME model. EMBEDDING_MODEL here MUST equal the constant in lib/embed.py.
 *
 * LAYER 0 STATUS: this module ships as correct parity code with NO call site
 * yet. Layer 0 wires retrieval only into the Python triage path (spec §6); the
 * web triage route still injects recency-ordered prior triages. Wire this in
 * when the web app needs query-side embeddings (e.g. when the web triage route
 * or a studio search adopts semantic retrieval).
 *
 * Placement note: this file lives in the repo-root `lib/` to sit beside
 * lib/embed.py as the shared retrieval home for the two-codebase repo. To import
 * it from the Next.js app (web/), it will need to be moved under web/lib/ or
 * reached via an added tsconfig path alias — Next cannot import from outside its
 * project root by default.
 *
 * Requires:
 *   - OPENAI_API_KEY in the environment
 *   - the `openai` npm package (install with: npm install openai)
 */

import OpenAI from "openai";

// ── Model lock — keep IN SYNC with lib/embed.py ─────────────────────────────────
export const EMBEDDING_MODEL = "text-embedding-3-small";
export const EMBEDDING_DIMENSIONS = 1536;

let _client: OpenAI | null = null;

function client(): OpenAI {
  if (_client) return _client;
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    throw new Error(
      "OPENAI_API_KEY is not set. Add it to the environment before embedding.",
    );
  }
  _client = new OpenAI({ apiKey });
  return _client;
}

/**
 * Embed a batch of strings. Returns one EMBEDDING_DIMENSIONS-length vector per
 * input, in the same order as the input array.
 */
export async function embedTexts(texts: string[]): Promise<number[][]> {
  if (texts.length === 0) return [];
  // The API rejects empty strings; substitute a single space to keep indices aligned.
  const cleaned = texts.map((t) => (t && t.trim() ? t : " "));
  const resp = await client().embeddings.create({
    model: EMBEDDING_MODEL,
    input: cleaned,
  });
  const sorted = [...resp.data].sort((a, b) => a.index - b.index);
  return sorted.map((item) => {
    if (item.embedding.length !== EMBEDDING_DIMENSIONS) {
      throw new Error(
        `Embedding dimension mismatch: got ${item.embedding.length}, ` +
          `expected ${EMBEDDING_DIMENSIONS} for model ${EMBEDDING_MODEL}`,
      );
    }
    return item.embedding;
  });
}

/** Embed a single string. Returns an EMBEDDING_DIMENSIONS-length vector. */
export async function embedText(text: string): Promise<number[]> {
  return (await embedTexts([text]))[0];
}

/**
 * Format a number[] as a pgvector literal ('[v0,v1,...]') for RPC through
 * PostgREST, which does not accept a raw JSON array for the vector type.
 */
export function toPgVector(vector: number[]): string {
  return "[" + vector.join(",") + "]";
}
